### PR TITLE
Introduce new protocol Record and RecordValue implementations with Jackson bindings

### LINKED ISSUE #7545: Immutables based protocol implementation
**Description**

In order to provide basic support for (de)serialization of `io.camunda.zeebe.protocol.Record` objects (and their values) to and from JSON, we will introduce a new module called `protocol-jackson`. This module will provide an implementation of the `protocol`'s `Record` and `RecordValue` (and children) interfaces with Jackson bindings, allowing for easy (de)serialization of these interfaces that is compatible with the JSON produced by `Record#toJson()`.

The implementation of these interfaces will be generated via the `Immutables` library.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR introduces a new module, `protocol-jackson`, which provides [Immutables](https://immutables.org) generated implementations for `Record<T>`, and all known `RecordValue` derived interfaces (e.g. `VariableRecordValue`, `TimerRecordValue`, etc.) at the time of writing. The primary goal here is to provide implementations which can be serialized and deserialized easily via Jackson, while being backwards compatible for deserialization with the JSON produced by the `Record#toJson` method and its current implementations in `protocol-impl`.

The actual generated implementations are not exposed via the public API to reduce the API surface, and reduce coupling with the code generator. Instead, we introduced new abstract types which implement the required interfaces (e.g. `AbstractRecord` implements `Record`, `AbstractVariableRecordValue` implements `VariableRecordValue`, etc.). These types are annotated with the appropriate Jackson annotations to ensure (de)serialization works as expected, including polymorphic (de)serialization via `Record#getValueType()`, e.g. the record's value and its concrete `Intent` type. What is exposed are these abstract types, and the generated builders.

The polymorphic (de)serialization is achieved by using Jackson's `TypeIdResolver` facilities. Implementations of this interface tell Jackson what is the concrete type to use when deserializing an object. So in our case, we have two: the `ValueTypeIdResolver` and the `IntentTypeIdResolver`, respectively for the record's `value` and the `intent`. Since in our protocol, we resolve types via the `valueType` property, we can use the `JsonTypeInfo` annotation to specify this. This means our `TypeIdResolver` implementations take in as input the actual `valueType` (already deserialized), and just need to return the correct Java class. This mapping is done in the package-private `ValueTypes` class, which is a thin wrapper around a constant `EnumMap`.

The abstract types are obviously exposed as they are what users are expected to tell Jackson is the type of their objects, either when creating readers/writers (e.g. `objectMapper.readerFor(AbstractRecord.class)` or `objectMapper.writerFor(AbstractRecord.class)`) or just reading/writing values (e.g. `objectMapper.readValue(myJson, AbstractRecord.class)` or `objectMapper.readValue(myJson, AbstractTimerRecordValue.class)`).

The builders are exposed primarily to allow users to build standalone records and values for unit testing. I'm not aware of any other production usage, so I'm happy to challenge this use case. It's also a bit of a long term one, as I would hope it can replace the `MockRecord` and eventually `CopiedRecord` and `CopiedRecords` functionalities. For example, a `new RecordBuilder().from(record).build()` would completely replace the `CopiedRecords` and `CopiedRecord` classes, as it would provide an immutable copy of the given record, regardless of its concrete implementation class. This may be a bit slower however and would need to be tested, especially regarding the serialization of these objects in the `ElasticsearchExporter`.

There are some known downsides to this PR. First of all, it slightly increases the maintenance burden when adding/removing record value types. Every time you add one, you need to also:

- Add a new abstract class for the value in the `protocol-jackson` implementation.
- Add a new mapping entry from the `ValueType` to its corresponding `Intent` and generated immutable class in `ValueTypes`. This is done for the polymorphic (de)serialization as mentioned above.

When removing a new value type, you'd have to reverse the above. This is a bit more than before, but it's considerably less than if we had to hand write all the new implementations, so I think it's acceptable. There are smoke tests - `IntentTypeIdResolverTest` and `ValueTypeIdResolverTest` - which ensure that we don't forget to map new `ValueType`.

Testing is done primarily via integration tests. As our main use case initially is being able to deserialize exported records, we have an integration test `ExporterSerializationIT` which starts a broker, runs the standard exporter workload (as defined in `ExporterIntegrationRule` of the `test` module), and fetches the JSON records from the `DebugHttpExporter`. They are deserialized, and then compared to the records from `RecordingExporter` to check that they are logically equivalent.

> I had some issues with logical equivalence of records: I've asked the AssertJ project about this and the Immutables project, I hope they can provide better insight. In the mean time, I went with a "hacky" solution. You can convert your `Record` objects to `JsonNode` - essentially an intermediate tree representation of the object. Since in JSON equivalence is basically just checking all the properties are the same, regardless of order and so on, this should work. It does seem strange, however.

## Related issues

closes #7545 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
Here's an overview of the various implementations I went through:

- [Generated code is completely exposed](https://github.com/camunda-cloud/zeebe/pull/7611/commits/0b58f19cc0c5837b1ad5185da9d3442d1b71a685). This initial implementation exposes the generated `Immutable*` classes and their builders, but hides the new `Abstract*` types. The idea there was that I wanted users to use _only_ the immutable classes and nothing else. This is my least favorite approach.
- [Builder is exposed via the Abstract* type and so are the Abstract types](https://github.com/camunda-cloud/zeebe/pull/7611/commits/bd36e6e59ae835f45959a691f26a3b28893dafcf).  The implementation classes (e.g. `ImmutableRecord`) are not exposed here, which means while the builder is public, it's not accessible. To circumvent this, you add a new static class called `Builder` in all your abstract types, which extends the immutable builder. So you then use `new AbstractRecord.Builder<>().from(record).build()`. It's an OK approach, but a little bit more verbose.
- [Builder is exposed as its own and so are the Abstract* types](https://github.com/camunda-cloud/zeebe/pull/7611/commits/4bc8960aad438f439479fcd2ba413483bf81f5a8). This is my favorite approach - you now only expose the builders, the immutable classes are nested and hidden inside. Your builders are final, and are essentially things like `RecordBuilder<>`, `VariableRecordBuilder`, etc., which you just instantiate as `new RecordBuilder<>()`, `new TimerRecordBuilder()`, and can use as `new RecordBuilder<>().from(record).build()`.

Note that when hiding the implementation classes, you now have to add to all abstract types the `JsonDeserialize(as = Immutable...`) annotation so Jackson knows how to actually deserialize this stuff. But I think that's pretty minimal boilerplate, as you can see from the code.

[Note also there's a way to completely hide the builder implementation as well](https://immutables.github.io/immutable.html#hide-implementation), however it's very verbose. You need to basically create an interface for the builder in your public API which already defines all methods you want accessible. I think that's too much in my opinion, and I'm fine with exposing the builder.

**Comment by npepinpe** [comment]
Also note that there's one downside to using the code generator. If you have a compile error somewhere else in the module, it may trip up the compiler and mean your classes will not be generated (since these are generated at compile time). So it will suddenly throw a whole bunch of compilation errors, when really there's only one that you care about, which generates a lot of noise. They're fairly easy to distinguish as they will all refer to `Immutable*`, but still...

**Comment by npepinpe** [comment]
I would not suggest reviewing by commits, apart from the commits I mentioned above (so you can see the differences if you're interested). It also looks like a lot (~1,100 insertions), but generally the abstract types (except `AbstractRecord`) are very similar, and the bulk of their files are just licensing and package imports. Feel free to check a couple, and their generated code, and then skip the rest. The interesting parts are: `AbstractRecord`, `ValueTypes`, the type ID resolvers (though only one of them is fine), and the tests.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractRecord.java
Comment (npepinpe) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractRecord.java:53]: :question: This was something I debated. Technically, as the object is immutable, there's no need to actually clone it. However, I think most people expect a new object (as in, by reference) when cloning, so this may break the contract. It's easy to replace with `new RecordBuilder<>().from(this).build()`, so let me know what you think.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java
Comment (npepinpe) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java:48]: :question: This keeps it aligned with how we handle records internally, but I don't know if that's a good idea or unexpected from an external user PoV?

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java
Comment (npepinpe) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java:9]: I opted against the abstract class and went with an interface. This is necessary to ensure the `toJson` property is auxilliary - that is, Immutables does not generate any `toJson` property, field, accessors, etc. So we mark it as `lazy`, which means it's computed once when called, but memoized. We could also mark it as `derived`, so it would be computed all the time, that's also fine. I pre-emptively mark them as deprecated as for this implementation, people should just use a configured `ObjectMapper` instead of this.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java
Comment (npepinpe) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java:22]: Happy to add comments to the settings and why I picked them, however they are very well documented so it shouldn't be too hard to figure it out. Let me know what you think.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java:9]: :question: Also, I debated making it public - technically it's exposed anyway since all the types implement it, so...meh. But it might be more explicit to not make it public if only so people aren't tempted to implement it :smile:

File: test/src/main/java/io/camunda/zeebe/test/exporter/ExporterIntegrationRule.java
Comment (npepinpe) [inline test/src/main/java/io/camunda/zeebe/test/exporter/ExporterIntegrationRule.java:238]: This was necessary to avoid having an explicit dependency on the `DebugHttpExporter` in the `protocol-jackson` module, which would cause me to have a dependency on the broker module, which is just silly. :clown_face:

**Comment by npepinpe** [comment]
SonarCloud complains of some duplication between `IntentTypeIdResolver` and `ValueTypeIdResolver`. I get that - I think it's the kind that's alright to have here instead of introducing some hierarchy, but let me know what you think.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java]: This was removed so... :man_shrugging:

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractJsonSerializable.java
Comment (korthout) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractJsonSerializable.java:29]: 🔧 I think that `JsonSerializable.toJson` should also be deprecated

**Review by korthout — COMMENTED** [review]
👏 Great work!

As expected, it took me some time to work through it. Perhaps not because of the amount of code, but rather the complexity of the work that resulted in those lines: the many interactions of the chosen libraries and how they may affect future maintenance as well as how users will likely use this new interface. I'm quite impressed at the depth of research you put into writing this and I think that you've already covered most questions I could ask.

I do have a couple comments/suggestions below, but nothing major. I will also reply to your earlier-asked questions, but note that also there nothing major will pop-up. Since, I do expect that you'll make some changes, I wait for those before giving my approval.

👍 To start off, regarding your 3 implementation ideas, I agree that the latter is the best option. It comes with some downsides which are all related to the generated class being referred to from the abstract class which it is generated from, but further extraction to interfaces complicates things a lot (adding maintenance costs and might be hard to achieve anyways). I also cannot imagine that we would move away from Immutables for this tiny part of our application, and even if we would, making it a breaking change out of it is not that bad.

💭 You mention backwards compatibility as a main goal. Now I think that this is all backwards compatible, but I was wondering whether we need to configure any checks for this. For example, does revapi automatically consider this a public api?

💡 Lastly, I think a README.md inside the `protocol-jackson` folder would be a great place to add a description of what this module is, how it's designed, what it can be used for and how. It could contain some of the choices you've described in this PR's body.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractRecord.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractRecord.java:53]: 👍 I'd leave it as it is

Reading the [Object.clone() docs](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#clone()), it's breaking some of the requirements, but it also states that these requirements are not absolute and depend on the context of the specific class.

The only problem I see with this approach is that it allows modification of the shared properties. For example, this.value refers to another object that may have its properties altered by calling methods on this.clone().value. Given the context we are in, it's safe to assume that T is also always an immutable object and cannot have its properties altered. I therefore think it's fine to break the `x.clone() != x` requirement.

The advantage is clearly in performance, and although I'm not one for premature optimization, I think we should just use the immutable context as an advantage here and just return `this`.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java:48]: ❓ What is the alternative? Throwing an exception or using null? I think what you've provided is the best out of those.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java]: ❓ This is no longer relevant right?

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java:22]: I don't think all options require explaining, but some definitely have been chosen due to some Zeebe intricacies. 

🔧  Please  add some explanation to: 

```
builderVisibility = BuilderVisibility.PUBLIC, // let's define why its public
implementationNestedInBuilder = true, // discussed above, could use a summary explanation here
defaultAsDefault = true, // unclear to me why we even need this
deepImmutablesDetection = true, // does quite a lot, so why exactly do we need it?
```

**Comment by korthout** [comment]
> SonarCloud complains of some duplication between `IntentTypeIdResolver` and `ValueTypeIdResolver`. I get that - I think it's the kind that's alright to have here instead of introducing some hierarchy, but let me know what you think.

I see no reason to avoid duplication there

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java:48]: :thought_balloon: It's a deserialization failure imo, so I would let it throw a `JsonProcessingException` error as an alternative. I'm actually not sure why we also have the `Unknown` intent in the first place :smile: 

Let's say this happens at runtime. How is your code supposed to react to this? We're now leaving it up to the user to check if it's the expected intention. Now let's say the user has some code like:

```java
final var record = deserialize(myJson);
if (record.getIntent() == Intent.UNKNOWN) {
  explodeOrDoSomethingWithIt();
}
```

How would they introspect what the original intent was though? I guess via the JSON...that's probably fine for most users I think.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypeIdResolver.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypeIdResolver.java:47]: From the method name :smile: I used `getTypeInfoOrNull` which indicates you may receive one or the other, whereas `getTypeInfo` will _always_ give you a type info or explode. What would you propose?

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypes.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypes.java:53]: This is more of a TODO for myself, I guess. I can remove it.

To clarify: type info can be extended to include the expected RecordValue interface (e.g. `VariableRecordValue.class`) and the mapping to the immutable equivalent (e.g. `ImmutableVariableRecordValue.class` via `new VariableRecordValueBuilder().from().build()`). This will give you a type safe way to easily map values to their immutable counter parts. You can then use this to easily recursively copy, say, a mutable `Record<?>`, converting it to an `ImmutableRecord<?>` _and_ also its `getValue()` property to the correct immutable value. But this isn't part of this PR, and while it makes it a bit tedious for the user to do the copy manually, I think that's OK for now.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java:31]: Ah, yes! I had this enabled previously for something else, but you're right - and I'd much rather use an interface than the abstract class. Good catch :muscle:

File: protocol-jackson/src/test/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolverTest.java
Comment (npepinpe) [reply protocol-jackson/src/test/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolverTest.java:48]: Why only some of them to var and not all? :upside_down_face:

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractJsonSerializable.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractJsonSerializable.java]: I wasn't yet sure about this, as that's definitely a breaking change as it's on the older module, and I wanted to think about it a bit more (also run it by menski).

**Comment by npepinpe** [comment]
Regarding your points:

- A `README.md` is definitely a must. I was planning on doing it and totally forgot.
- Right now `revapi` does not run on this module. I was thinking of actually not having any checks for the first version/iteration - we can test it out with Optimize/Operate/Tasklist, get some feedback, and once we have a better idea we can add checks. So what I'm saying is, eventually I want backwards compatibility with this, but for now I would consider it an incubating feature which we can play around a bit for at least one version more :smile:

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java:48]: I see your point, but throwing it in the `mapValueTypeToIntentClass` scope does not give access to the json directly right? Or does jackson decorate the exception with useful data if you throw `JsonProcessingException` inside that method? If that's the case, then let's fail early. Otherwise, let's fail late and allow the user to discover the context

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypeIdResolver.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypeIdResolver.java:47]: 💭 philosophically, an `IllegalArgumentException` should only be thrown when the method was called with an argument you can't accept. The `getTypeInfo` method can easily accept any `ValueType`, and just return a value if it finds something, or nothing if it doesn't (i.e. `Optional.empty` or `null`). You don't really need 2 methods for that. It's only on the caller's side of the method (in the resolver) where you want to make the distinction on what to do if no type info was found: e.g. fail by throwing, recovering, etc.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypes.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ValueTypes.java:53]: Thanks for explaining. I'd rather have it left out of the javadoc in that case. Feel free to turn it into a `//todo(@npepinpe):...` comment though.

File: protocol-jackson/src/test/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolverTest.java
Comment (korthout) [reply protocol-jackson/src/test/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolverTest.java]: `var` was not intended to replace all local variable type declarations. The original JEP came with a [style guide](https://openjdk.java.net/projects/amber/LVTIstyle.html) that proposed to use it in specific cases where the type is already clear from the rest of the line or statement, or where you otherwise don't care about the type. For example, calling a constructor, or a method who's name clearly shows what type it returns. In some cases, understandability would be reduced by inferring the type.

I think the `DefaultDeserializationContext` and `JavaType` could offer some additional insight into what this code does, but I don't care too much for it. Feel free to change it how you think it reads best.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractJsonSerializable.java
Comment (korthout) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractJsonSerializable.java]: ❓ Deprecation is not a breaking change right? It's simply stating that in the next major version, this method may be removed.

👍 It's good to give this more consideration though

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java
Comment (korthout) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/LazyJsonSerializable.java:22]: I don't really see how this default implementation is lazy. 

🔧 Please either keep track of the first result and return it on each consecutive call; or rename the interface to something like `SimpleJsonSerializable` or `DefaultJsonSerializable` or something else.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java
Comment (korthout) [inline protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java]: `defaultAsDefault` is false by default 😄 

🔧 To avoid confusion, just remove it altogether.

**Review by korthout — COMMENTED** [review]
Thanks for the quick replies. Here's a quick review of the recent changes.

Looking forward to the README.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/ZeebeStyle.java]: Already done ;)

File: protocol-jackson/README.md
Comment (korthout) [inline protocol-jackson/README.md:96]: 💭 Could be nice to add here that this is an alternative to copy and provides complete deep copying. However, given the immutable nature of the generated classes, copy can also be used directly for faster copying.

File: protocol-jackson/README.md
Comment (korthout) [inline protocol-jackson/README.md:33]: 🙃  Why are you consistently not using a space around the equals char?

File: protocol-jackson/README.md
Comment (korthout) [inline protocol-jackson/README.md:108]: 🔧  Let's refer to this header in the test cases that check that a `ValueType` is unmapped.

**Review by korthout — APPROVED** [review]
LGTM 🔥 

That's a well written README. I imagine it will help future devs working in many different teams: e.g. Zeebe/Operate/Tasklist/Optimize/etc. Thanks for this. 

Please consider the following suggestions/questions:

**Comment by korthout** [comment]
@npepinpe ❓ Should this module be build using jdk 1.8? Also, should the module be added to the bom?

**Comment by npepinpe** [comment]
Ah yes, I forgot - definitely should be added to the BOM. Regarding 1.8, that's a good question. At the moment, the only one with a hard requirement to 1.8 is the client module (and by definition, the modules which it requires). I would leave it as is for now, that's more of a product decision.

File: protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java
Comment (npepinpe) [reply protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/IntentTypeIdResolver.java:48]: Let's leave it as is then.

File: protocol-jackson/README.md
Comment (npepinpe) [reply protocol-jackson/README.md:96]: It doesn't provide deep copying actually. If you got that from this, then I need to rewrite it :smile:

File: protocol-jackson/README.md
Comment (npepinpe) [reply protocol-jackson/README.md:33]: That's the auto formatter :smile:

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**
